New dimensiondata firewall

tintoy opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### COMPONENT NAME

ansible-module-extras/cloud/dimensiondata/dimensiondata_firewall.py
##### ANSIBLE VERSION

```
ansible 2.2.0
```
##### SUMMARY

This is Ansible module support for apache-libcloud functions.
Specifically, this code supports Ansible invocations of DimensionData's 
firewall functionality.

This is to support a new capability within ansible, so there is no "BEFORE".
Note that this routine has an "action verb" which allows multiple different
actions to be performed on a firewall.  Below shows the ansible call to
create a firewall, along with its output.

```
    - name: Create MCP 2.0 firewall rule
      dimensiondata_firewall:
        region: "{{dimension_data_region}}"
        location: "{{dimension_data_location}}"
        network_domain: "{{dimension_data_network_domain}}"
        name: firewall_ansible_test
        state: enabled
        action: ACCEPT_DECISIVELY
        ip_version: IPv4
        protocol: TCP
        source: ANY
        source_port: ANY
        destination: 168.128.28.100
        destination_port: 5555
        position: LAST

    TASK [Create MCP 2.0 firewall rule] ********************************************
    changed: [127.0.0.1] => {"changed": true, "firewall_rule": {"action": "ACCEPT_DECISIVELY", "enabled": "true", "id": "3a456e4d-69e1-4650-b9ba-3d1ce42bdfbb", "ip_version": "IPv4", "location": {"country": "US", "id": "NA12", "name": "US - West - MCP 2.0"}, "name": "firewall_ansible_test", "protocol": "TCP", "status": null}, "msg": "Firewall rule created successfully."}

```
###### REQUIRES

https://github.com/ansible/ansible/pull/17604
https://github.com/apache/libcloud/pull/858

Commit:
Add dimensiondata_firewall module

jr2730 committed:
fixed error message formatting

tintoy commented:
Migrated from https://github.com/ansible/ansible-modules-extras/pull/2983 by tintoy (not original author)

tintoy commented:
Rebasing against `devel`.

tintoy committed:
Tidy up code based on feedback from previous PRs.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+description:
+    - Create, update, and delete MCP 2.0 firewall rules.
+    - Requires MCP 2.0.
+version_added: "2.2"
```
`"2.3"`

tintoy replied:
Sorry, I really should have caught this one. I'll make a note to check it when porting the other outstanding PRs.

tintoy replied:
Done.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+DOCUMENTATION = '''
+---
+module: dimensiondata_firewall
+short_description: Create, update, and delete MCP 2.0 firewall rules.
```
No full stop needed on `short_description`.

All other `description:` fields should be full sentences.

tintoy replied:
Done.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+    description:
+      - The target region.
+      - Valid regions are defined in Apache libcloud project [libcloud/common/dimensiondata.py]
+      - Regions are also listed in https://libcloud.readthedocs.io/en/latest/compute/drivers/dimensiondata.html
```
Changing this to the following will give you a clickable URL
`U(https://libcloud.readthedocs.io/en/latest/compute/drivers/dimensiondata.html)`

tintoy replied:
Done. Thanks for the tip!

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+      - The target region.
+      - Valid regions are defined in Apache libcloud project [libcloud/common/dimensiondata.py]
+      - Regions are also listed in https://libcloud.readthedocs.io/en/latest/compute/drivers/dimensiondata.html
+      - Note that the default value "na" stands for "North America".
```
+1 to adding that note :)

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+  mcp_user:
+    description:
+      - The username used to authenticate to the CloudControl API.
+      - If not specified, will fall back to MCP_USER from environment variable or  ~/.dimensiondata.
```
`If not specified, will fall back to the C(MCP_USER) environment variable or credentials specified in  C(~/.dimensiondata).`

tintoy replied:
Done.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+  mcp_password:
+    description:
+      - The password used to authenticate to the CloudControl API.
+      - If not specified, will fall back to MCP_PASSWORD from environment variable or  ~/.dimensiondata.
```
Please update formatting as per `mcp_user`

tintoy replied:
Done.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+    description:
+      - The password used to authenticate to the CloudControl API.
+      - If not specified, will fall back to MCP_PASSWORD from environment variable or  ~/.dimensiondata.
+      - Required if mcp_user is specified.
```
This needs to be enforced in code via `required_together`, see other modules for examples.

tintoy replied:
Done (I think - sanity check appreciated).

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+    description:
+      - Action to take when rule matched.
+      - ACCEPT_DECISIVELY or DROP
+    choices: [ACCEPT_DECISIVELY, DROP]
```
`choices: [ "ACCEPT_DECISIVELY", "DROP" ]`

tintoy replied:
How about for `default`? Same deal?

tintoy replied:
Did it for all of them - hope that's correct.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+            sample: enabled
+'''
+
+from ansible.module_utils.basic import *
```
Please only import what you need, rather than `*`.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+'''
+
+from ansible.module_utils.basic import *
+from ansible.module_utils.dimensiondata import *
```
Please only import what you need, rather than `*`.

tintoy replied:
Done.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+        )
+
+
+main()
```
```
if __name__ == '__main__':
    main()
```

tintoy replied:
Yeah, I was wondering... I've picked up these PRs from someone who is no longer with us, so I'm still becoming familiar with how this stuff is supposed to work. Sorry about that, will fix :)

tintoy replied:
Done.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,510 @@
+        else:
+            # Delete rule
+            delete_firewall_rule(module, driver, existing_rule)
+    else:
```
You shouldn't need this, it's enforced by Ansible as you've already specified
`choices=['present', 'absent', 'enabled', 'disabled']),`

tintoy replied:
Done (I think - if you have a moment I'd appreciate a sanity check that I'm now doing it right).

gundalow reviewed (changes requested):
needs_revision

Please also update the CHANGELOG.md in the root of the repo to add this new module under the 2.3 section

tintoy commented:
Hi @gundalow, thanks for the feedback and tips - I'll get right onto this tomorrow :)

tintoy committed:
Tidy up code based on feedback from @gundalow.

- Simplify imports
- Tidy up module documentation
- Remove explicit checks for invalid values; module validation should catch them.
- Only call main() when module is run directly (i.e. __name__ == '__main__').

gundalow commented:
Looking a lot better. Feel free to stop by `#ansible-devel` on Freenode IRC if you'd like more interactive feedback :)

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,509 @@
+  action:
+    description:
+      - Action to take when rule matched.
+      - ACCEPT_DECISIVELY or DROP
```
You can delete this line.

gundalow commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,509 @@
+            verify_ssl_cert=dict(required=False, default=True, type='bool')
+        ),
+        required_together=[
+            ['mcp_user', 'mcp_password']
```
Looks good. You can test this by specifying one but not the other in a task and ensuring you get a sensible error

tintoy committed:
Remove redundant specifications of choices in module documentation (ansible/ansible#20583).

gundalow commented:
Let me know once you are happy and I'll review again. 

If you let me know the GitHub IDs of some coworkers I can add them as maintainers of this module, this means they will be notified if bugs are raised against any of these modules.

tintoy commented:
It's just me for now - I'm happy to be listed as the go-to person for bug reports. I'll see if anyone else would like to get involved though :)

tintoy commented:
@gundalow I'm seeing some odd behaviour when trying to run ansible-playbook after using `source hacking/env-setup`.

We depend on [module_utils/dimensiondata.py](https://github.com/tintoy/ansible/blob/ansible-modules-extras/pull/2983/lib/ansible/module_utils/dimensiondata.py#L25). But when I run `ansible-playbook` to test this new module, I get:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: ImportError: No module named configparser
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/yv/x0nzk5f901g4fywjhzdy18qr0000gn/T/ansible_12CqIc/ansible_module_dimensiondata_network.py\", line 158, in <module>\n    from ansible.module_utils.dimensiondata import get_credentials, DimensionDataAPIException, LibcloudNotFound\n  File \"/var/folders/yv/x0nzk5f901g4fywjhzdy18qr0000gn/T/ansible_12CqIc/ansible_modlib.zip/ansible/module_utils/dimensiondata.py\", line 25, in <module>\nImportError: No module named configparser\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 0}
```

Anything obvious that I might be missing? I can see that `six.py` does, indeed, include a mapping for ConfigParser.

tintoy commented:
(sorry about the formatting)

tintoy commented:
@gundalow interestingly, when I switch back to:

```python
from configparser import ConfigParser
```

instead of:

```python
from ansible.module_utils.six.moves.configparser import ConfigParser
```

(which I was asked to use instead) then everything works correctly. Is my local setup perhaps missing something required in order for the `configparser` module to show up under `ansible.module_utils.six.moves`? I've tried this in a fresh venv and the behaviour is the same (works with the first style of import, but not the second).

tintoy commented:
Ok, so this also works:

```python
from ansible.compat.six.moves import configparser
from configparser import ConfigParser
```

Is that acceptable, @gundalow?

abadger commented:
```
from configparser import ConfigParser
```

Will work on python3.x and boxes that have the python-future 3rd party library installed.  It doesn't work out of the box on python2.7 or python2.6 (From a centos7 machine:
```
[toshio@peven ~]$ python2
Python 2.7.5 (default, Nov  6 2016, 00:28:07) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-11)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from configparser import ConfigParser
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: No module named configparser
```

This works on boxes which have ansible installed:
```
from ansible.compat.six.moves import configparser
```
But it won't work when used against a box which doesn't have ansible installed (only things in ansible.module_utils are added to the zip file along with the module and pushed across the wire to the remote machine).

I can reproduce your issue, though, which only seems to affect configparser (not other things like ```from ansible.module_utils.six.moves.urllib import parse)```).  It could be a bug in the older version of six that we currently have to bundle in module_utils (because of python-2.4 compatibility).  I'll look through the six changelog to see if something changed there.  In the meantime, does it work if you do it this way?

``` python
from ansible.module_utils.six.moves import configparser
ConfigParser = configparser.ConfigParser
```

abadger commented:
Yeah, the six library was fixed here: https://github.com/benjaminp/six/commit/8d7f9208af650d895d5fb08f400f614638ec66f4 which seems to have been the same release where python-2.4 support was dropped.  So we'll need to use the workaround for the Ansible-2.3 release.  Once we fork devel for the Ansible-2.4 release we can update the version of six in module_utils and switch back to ```from ansible.module_utils.six.moves.configparser import ConfigParser``` at our leisure.

Note -- I'm looking at how the module_util uses configparser right now.  I think it's slightly wrong right now.  Here's how it is now:
``` python
from ansible.module_utils.six.moves.configparser import ConfigParser
[...]
        config = ConfigParser.RawConfigParser()
[...]
        except (ConfigParser.NoSectionError, ConfigParser.NoOptionError):
```

Here's how I think it should be (which won't even need our workaround):
``` python
from ansible.module_utils.six.moves import configparser
[...]
        config = configparser.RawConfigParser()
[...]
        except (configparser.NoSectionError, configparser.NoOptionError):
```

tintoy commented:
Yep, that makes much more sense - thanks! I'll make those changes shortly.

tintoy committed:
More-compatible import of ConfigParser.

Also, port module_utils changes for MCP version detection from PR ansible/ansible#20321.

tintoy commented:
Yep, thanks @abadger, that works :)

mattclay commented on lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py:
```diff
@@ -0,0 +1,506 @@
+#   - Jay Riddell    <Jay.Riddell@dimensiondata.com>
+#   - Adam Friedman  <tintoy@tintoy.io>
+#
+
```
CI failure due to missing module metadata:

```
2017-02-03 18:02:56 ============================================================================
2017-02-03 18:02:56 lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py
2017-02-03 18:02:56 ============================================================================
2017-02-03 18:02:56 ERROR: No ANSIBLE_METADATA provided
```

tintoy replied:
Oops! No idea why that went missing - I'll add it now.

tintoy replied:
Actually, neither of the `dimensiondata` modules currently has `ANSIBLE_METADATA`. Are you happy for me to add it to both of them for this PR?

mattclay replied:
Actually, the `dimensiondata_network` module already has metadata:

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/dimensiondata/dimensiondata_network.py#L25

mattclay replied:
You'll need to rebase your branch to pull in the metadata and resolve the merge conflict.

tintoy replied:
Hmm, sounds like the PR branch is out-of-sync with `devel`. Time to rebase, perhaps?

tintoy replied:
Snap! :)

Ok, will do.

tintoy replied:
Righto - have rebased - should be good now.

tintoy committed:
Add ANSIBLE_METADATA to both dimensiondata_network and dimensiondata_firewall modules.

tintoy commented:
Hi - I'm sorry to mess you guys around, but after reviewing the remaining pull requests, I've decided to bite the bullet and refactor the `dimensiondata_network` module and its dependencies in `module_utils` (currently, there is quite a lot of duplicated code in each module). Ideally, this won't mean that any of the work-to-date will be wasted but will instead make the reviews much simpler (since there will be more shared code). I'll open a PR shortly with the changes to `dimensiondata_network` and `module_utils` for your review (hopefully I'm moving in the right direction).

gundalow commented:
@tintoy I don't think anyone will mind you refactoring this into a cleaner set of modules :) 

Once the new PRs are raised just add a comment which links to the new PRs here.

tintoy commented:
Good to hear :-)

I'm starting with ansible/ansible#21043.
